File: tests/test-metrics.js
/**
 * Test suite for metrics.js
 * Tests metrics calculation functionality
 */

const tests = [];
let passedTests = 0;
let failedTests = 0;

function test(name, fn) {
    tests.push({ name, fn });
}

function assertEquals(actual, expected, message) {
    if (Math.abs(actual - expected) > 0.001) { // Allow small floating point differences
        throw new Error(`${message}\nExpected: ${expected}\nActual: ${actual}`);
    }
}

function assertTrue(condition, message) {
    if (!condition) {
        throw new Error(message);
    }
}

function runTests() {
    console.log('Running metrics.js tests...\n');

    for (const { name, fn } of tests) {
        try {
            fn();
            console.log(`✓ ${name}`);
            passedTests++;
        } catch (error) {
            console.error(`✗ ${name}`);
            console.error(`  ${error.message}`);
            failedTests++;
        }
    }

    console.log(`\n${passedTests} passed, ${failedTests} failed`);
    return failedTests === 0;
}

// Tests

test('Metrics: perfect match', () => {
    const matches = [
        ['hello', 'hello', 0, 'exact'],
        ['world', 'world', 0, 'exact']
    ];

    const metrics = calculateMetrics(matches);

    assertEquals(metrics.precision, 1.0, 'Precision should be 1.0');
    assertEquals(metrics.recall, 1.0, 'Recall should be 1.0');
    assertEquals(metrics.f1_score, 1.0, 'F1 score should be 1.0');
    assertEquals(metrics.avg_crr, 1.0, 'CRR should be 1.0');
    assertEquals(metrics.exact_matches, 2, 'Should have 2 exact matches');
});

test('Metrics: no matches', () => {
    const matches = [
        ['hello', null, null, 'gt_only'],
        ['world', null, null, 'gt_only'],
        [null, 'foo', null, 'ocr_only'],
        [null, 'bar', null, 'ocr_only']
    ];

    const metrics = calculateMetrics(matches);

    assertEquals(metrics.precision, 0.0, 'Precision should be 0.0');
    assertEquals(metrics.recall, 0.0, 'Recall should be 0.0');
    assertEquals(metrics.f1_score, 0.0, 'F1 score should be 0.0');
    assertEquals(metrics.exact_matches, 0, 'Should have 0 exact matches');
});

test('Metrics: half precision', () => {
    const matches = [
        ['hello', 'hello', 0, 'exact'],
        [null, 'world', null, 'ocr_only'] // OCR hallucinated this
    ];

    const metrics = calculateMetrics(matches);

    assertEquals(metrics.precision, 0.5, 'Precision should be 0.5 (1/2)');
    assertEquals(metrics.recall, 1.0, 'Recall should be 1.0 (1/1)');
});

test('Metrics: half recall', () => {
    const matches = [
        ['hello', 'hello', 0, 'exact'],
        ['world', null, null, 'gt_only'] // OCR missed this
    ];

    const metrics = calculateMetrics(matches);

    assertEquals(metrics.precision, 1.0, 'Precision should be 1.0 (1/1)');
    assertEquals(metrics.recall, 0.5, 'Recall should be 0.5 (1/2)');
});

test('Metrics: F1 score calculation', () => {
    const matches = [
        ['the', 'the', 0, 'exact'],
        ['quick', 'quick', 0, 'exact'],
        ['brown', null, null, 'gt_only'],
        ['fox', null, null, 'gt_only'],
        [null, 'quik', null, 'ocr_only']
    ];

    const metrics = calculateMetrics(matches);

    // Precision: 2/3 = 0.667
    // Recall: 2/4 = 0.5
    // F1: 2 * (0.667 * 0.5) / (0.667 + 0.5) = 0.571
    assertEquals(metrics.precision, 2/3, 'Precision should be 2/3');
    assertEquals(metrics.recall, 0.5, 'Recall should be 0.5');
    assertEquals(metrics.f1_score, 0.571, 'F1 score should be ~0.571');
});

test('Metrics: CRR with exact match only', () => {
    const matches = [
        ['hello', 'hello', 0, 'exact'] // 0 character errors
    ];

    const metrics = calculateMetrics(matches);

    // Total chars: 5 (hello)
    // Errors: 0
    // CRR: 1 - (0/5) = 1.0
    assertEquals(metrics.avg_crr, 1.0, 'CRR should be 1.0 for exact match');
});

test('Metrics: CRR with unmatched words', () => {
    const matches = [
        ['hello', 'hello', 0, 'exact'], // 0 errors
        ['world', null, null, 'gt_only'], // 5 chars = 5 errors
        [null, 'test', null, 'ocr_only'] // 4 chars = 4 errors
    ];

    const metrics = calculateMetrics(matches);

    // Total GT chars: 5 + 5 = 10
    // Total errors: 0 + 5 + 4 = 9
    // CRR: 1 - (9/10) = 0.1
    assertEquals(metrics.avg_crr, 0.1, 'CRR should be 0.1');
});

test('Metrics: invoice example', () => {
    // Ground truth: "Invoice Number: INV-2024-001"
    // OCR output: "Invoice Number: INV-2O24-001" (O instead of 0)
    // After normalization: invoice, number, inv, 2024, 001 vs invoice, number, inv, 2o24, 001
    const matches = [
        ['invoice', 'invoice', 0, 'exact'],
        ['number', 'number', 0, 'exact'],
        ['inv', 'inv', 0, 'exact'],
        ['2024', null, null, 'gt_only'],
        ['001', '001', 0, 'exact'],
        [null, '2o24', null, 'ocr_only']
    ];

    const metrics = calculateMetrics(matches);

    // Precision: 4/5 = 0.8
    // Recall: 4/5 = 0.8
    // F1: 2 * (0.8 * 0.8) / (0.8 + 0.8) = 0.8
    assertEquals(metrics.precision, 0.8, 'Precision should be 0.8');
    assertEquals(metrics.recall, 0.8, 'Recall should be 0.8');
    assertEquals(metrics.f1_score, 0.8, 'F1 score should be 0.8');
});

test('Metrics: counts are correct', () => {
    const matches = [
        ['a', 'a', 0, 'exact'],
        ['b', 'b', 0, 'exact'],
        ['c', null, null, 'gt_only'],
        ['d', null, null, 'gt_only'],
        [null, 'e', null, 'ocr_only']
    ];

    const metrics = calculateMetrics(matches);

    assertEquals(metrics.exact_matches, 2, 'Should have 2 exact matches');
    assertEquals(metrics.total_gt_words, 4, 'Should have 4 GT words');
    assertEquals(metrics.total_ocr_words, 3, 'Should have 3 OCR words');
    assertEquals(metrics.unmatched_gt, 2, 'Should have 2 unmatched GT words');
    assertEquals(metrics.unmatched_ocr, 1, 'Should have 1 unmatched OCR word');
});

test('Metrics: CRR comprehensive calculation', () => {
    // Testing all CRR scenarios:
    // - Exact matches contribute 0 errors
    // - GT-only words: all chars are errors
    // - OCR-only words: all chars are errors
    const matches = [
        ['the', 'the', 0, 'exact'],      // 3 chars, 0 errors
        ['quick', 'quick', 0, 'exact'],  // 5 chars, 0 errors
        ['brown', null, null, 'gt_only'], // 5 chars, 5 errors (missed by OCR)
        ['fox', null, null, 'gt_only'],   // 3 chars, 3 errors (missed by OCR)
        [null, 'quik', null, 'ocr_only']  // 4 chars, 4 errors (hallucinated)
    ];

    const metrics = calculateMetrics(matches);

    // Total GT chars: 3 (the) + 5 (quick) + 5 (brown) + 3 (fox) = 16
    // Total errors: 0 (the) + 0 (quick) + 5 (brown) + 3 (fox) + 4 (quik) = 12
    // CRR: 1 - (12/16) = 0.25
    assertEquals(metrics.avg_crr, 0.25, 'CRR should be 0.25 (25%)');
});

test('Metrics: empty input', () => {
    const matches = [];

    const metrics = calculateMetrics(matches);

    assertEquals(metrics.precision, 0.0, 'Precision should be 0.0 for empty input');
    assertEquals(metrics.recall, 0.0, 'Recall should be 0.0 for empty input');
    assertEquals(metrics.f1_score, 0.0, 'F1 score should be 0.0 for empty input');
    assertEquals(metrics.avg_crr, 0.0, 'CRR should be 0.0 for empty input');
});

// Run tests if in browser
if (typeof window !== 'undefined') {
    window.addEventListener('DOMContentLoaded', () => {
        const success = runTests();
        if (!success) {
            console.error('Some tests failed!');
        }
    });
}

// Export for Node.js testing
if (typeof module !== 'undefined' && module.exports) {
    module.exports = { test, runTests, assertEquals, assertTrue };
}

File: tests/test_metrics.py
"""
Unit tests for core/metrics.py
Tests precision, recall, F1 score, and CRR calculations
"""

import unittest
from core.metrics import calculate_metrics, format_metrics_for_display


class TestCalculateMetrics(unittest.TestCase):
    """Test metrics calculation"""

    def test_all_exact_matches(self):
        """All words match exactly - perfect score"""
        matches = [
            ('hello', 'hello', 0, 'exact'),
            ('world', 'world', 0, 'exact')
        ]

        metrics = calculate_metrics(matches)

        self.assertEqual(metrics['precision'], 1.0)  # 2/2
        self.assertEqual(metrics['recall'], 1.0)  # 2/2
        self.assertEqual(metrics['f1_score'], 1.0)
        self.assertEqual(metrics['avg_crr'], 1.0)  # No errors
        self.assertEqual(metrics['exact_matches'], 2)
        self.assertEqual(metrics['fuzzy_matches'], 0)
        self.assertEqual(metrics['total_gt_words'], 2)
        self.assertEqual(metrics['total_ocr_words'], 2)

    def test_no_matches(self):
        """No matching words - worst case"""
        matches = [
            ('hello', None, None, 'gt_only'),
            ('world', None, None, 'gt_only'),
            (None, 'foo', None, 'ocr_only'),
            (None, 'bar', None, 'ocr_only')
        ]

        metrics = calculate_metrics(matches)

        self.assertEqual(metrics['precision'], 0.0)  # 0/2
        self.assertEqual(metrics['recall'], 0.0)  # 0/2
        self.assertEqual(metrics['f1_score'], 0.0)
        self.assertEqual(metrics['exact_matches'], 0)
        self.assertEqual(metrics['unmatched_gt'], 2)
        self.assertEqual(metrics['unmatched_ocr'], 2)

    def test_precision_calculation(self):
        """Test precision: exact_matches / total_ocr_words"""
        # GT: "hello world test"
        # OCR: "hello world foo"
        matches = [
            ('hello', 'hello', 0, 'exact'),
            ('world', 'world', 0, 'exact'),
            ('test', None, None, 'gt_only'),
            (None, 'foo', None, 'ocr_only')
        ]

        metrics = calculate_metrics(matches)

        # Precision: 2 exact matches / 3 total OCR words = 2/3
        self.assertAlmostEqual(metrics['precision'], 0.6667, places=4)

    def test_recall_calculation(self):
        """Test recall: exact_matches / total_gt_words"""
        # GT: "hello world test"
        # OCR: "hello world"
        matches = [
            ('hello', 'hello', 0, 'exact'),
            ('world', 'world', 0, 'exact'),
            ('test', None, None, 'gt_only')
        ]

        metrics = calculate_metrics(matches)

        # Recall: 2 exact matches / 3 total GT words = 2/3
        self.assertAlmostEqual(metrics['recall'], 0.6667, places=4)

    def test_f1_score_calculation(self):
        """Test F1 score: harmonic mean of precision and recall"""
        # Precision = 2/3, Recall = 2/4 = 0.5
        # F1 = 2 * (0.6667 * 0.5) / (0.6667 + 0.5) = 2 * 0.3333 / 1.1667 = 0.5714
        matches = [
            ('hello', 'hello', 0, 'exact'),
            ('world', 'world', 0, 'exact'),
            ('quick', None, None, 'gt_only'),
            ('fox', None, None, 'gt_only'),
            (None, 'test', None, 'ocr_only')
        ]

        metrics = calculate_metrics(matches)

        precision = 2 / 3  # 0.6667
        recall = 2 / 4  # 0.5
        expected_f1 = 2 * (precision * recall) / (precision + recall)

        self.assertAlmostEqual(metrics['f1_score'], expected_f1, places=4)

    def test_crr_only_exact_matches(self):
        """CRR with only exact matches"""
        # All exact, no errors
        matches = [
            ('hello', 'hello', 0, 'exact'),
            ('world', 'world', 0, 'exact')
        ]

        metrics = calculate_metrics(matches)

        # CRR = 1 - (0 errors / 10 chars) = 1.0
        self.assertEqual(metrics['avg_crr'], 1.0)

    def test_crr_with_unmatched_gt_words(self):
        """CRR counts unmatched GT words as errors"""
        # GT: "hello world" (10 chars)
        # OCR: "hello" (5 chars)
        # Unmatched: "world" (5 chars)
        matches = [
            ('hello', 'hello', 0, 'exact'),
            ('world', None, None, 'gt_only')
        ]

        metrics = calculate_metrics(matches)

        # CRR = 1 - (5 errors / 10 total chars) = 0.5
        self.assertEqual(metrics['avg_crr'], 0.5)

    def test_crr_with_unmatched_ocr_words(self):
        """CRR counts unmatched OCR words as errors"""
        # GT: "hello" (5 chars)
        # OCR: "hello world" (10 chars)
        # Unmatched: "world" (5 chars - hallucinated)
        matches = [
            ('hello', 'hello', 0, 'exact'),
            (None, 'world', None, 'ocr_only')
        ]

        metrics = calculate_metrics(matches)

        # Total errors = 5 (unmatched OCR chars)
        # Total GT chars = 5
        # CRR = 1 - (5/5) = 0.0
        self.assertEqual(metrics['avg_crr'], 0.0)

    def test_crr_comprehensive_example(self):
        """Comprehensive CRR test with README example"""
        # GT: "The quick brown fox" (16 chars)
        # OCR: "The quik brown" (13 chars)
        # Exact: "The" (0 errors), "brown" (0 errors)
        # Unmatched GT: "quick" (5 chars), "fox" (3 chars) = 8 errors
        # Unmatched OCR: "quik" (4 chars) = 4 errors
        # Total errors: 12
        # CRR = 1 - (12/16) = 0.25

        matches = [
            ('The', 'The', 0, 'exact'),
            ('brown', 'brown', 0, 'exact'),
            ('quick', None, None, 'gt_only'),
            ('fox', None, None, 'gt_only'),
            (None, 'quik', None, 'ocr_only')
        ]

        metrics = calculate_metrics(matches)

        # Total errors = 0 + 0 + 5 + 3 + 4 = 12
        # Total GT chars = 3 + 5 + 5 + 3 = 16
        # CRR = 1 - 12/16 = 0.25
        self.assertEqual(metrics['avg_crr'], 0.25)

    def test_fuzzy_matches_dont_count_for_precision_recall(self):
        """Fuzzy matches should NOT count toward precision/recall"""
        matches = [
            ('hello', 'hello', 0, 'exact'),
            ('world', 'wrld', 1, 'fuzzy')  # Fuzzy match
        ]

        metrics = calculate_metrics(matches)

        # Only exact match counts
        self.assertEqual(metrics['precision'], 0.5)  # 1/2
        self.assertEqual(metrics['recall'], 0.5)  # 1/2
        self.assertEqual(metrics['exact_matches'], 1)
        self.assertEqual(metrics['fuzzy_matches'], 1)

    def test_fuzzy_matches_count_for_crr(self):
        """Fuzzy matches DO count toward CRR"""
        # GT: "hello" (5 chars)
        # OCR: "helo" (4 chars)
        # Fuzzy match with edit distance 1
        matches = [
            ('hello', 'helo', 1, 'fuzzy')
        ]

        metrics = calculate_metrics(matches)

        # CRR = 1 - (1 error / 5 chars) = 0.8
        self.assertEqual(metrics['avg_crr'], 0.8)

    def test_empty_matches(self):
        """Handle empty match list"""
        matches = []

        metrics = calculate_metrics(matches)

        self.assertEqual(metrics['precision'], 0.0)
        self.assertEqual(metrics['recall'], 0.0)
        self.assertEqual(metrics['f1_score'], 0.0)
        self.assertEqual(metrics['avg_crr'], 0.0)
        self.assertEqual(metrics['exact_matches'], 0)

    def test_zero_division_handling(self):
        """Handle zero division cases gracefully"""
        # Only GT words, no OCR
        matches = [
            ('hello', None, None, 'gt_only')
        ]

        metrics = calculate_metrics(matches)

        # Precision should be 0 (0 exact / 0 OCR words)
        self.assertEqual(metrics['precision'], 0.0)
        # Recall should be 0 (0 exact / 1 GT word)
        self.assertEqual(metrics['recall'], 0.0)

    def test_counts_are_correct(self):
        """Verify all count fields are correct"""
        matches = [
            ('a', 'a', 0, 'exact'),
            ('b', 'b', 0, 'exact'),
            ('c', 'd', 1, 'fuzzy'),
            ('e', None, None, 'gt_only'),
            (None, 'f', None, 'ocr_only')
        ]

        metrics = calculate_metrics(matches)

        self.assertEqual(metrics['exact_matches'], 2)
        self.assertEqual(metrics['fuzzy_matches'], 1)
        self.assertEqual(metrics['total_gt_words'], 4)  # a, b, c, e
        self.assertEqual(metrics['total_ocr_words'], 4)  # a, b, d, f
        self.assertEqual(metrics['unmatched_gt'], 1)  # e
        self.assertEqual(metrics['unmatched_ocr'], 1)  # f


class TestFormatMetricsForDisplay(unittest.TestCase):
    """Test metrics formatting for display"""

    def test_format_percentages(self):
        """Format metrics as percentages"""
        metrics = {
            'precision': 0.6667,
            'recall': 0.5,
            'avg_crr': 0.75,
            'f1_score': 0.5714,
            'exact_matches': 2,
            'fuzzy_matches': 1,
            'total_gt_words': 3,
            'total_ocr_words': 3,
            'unmatched_gt': 1,
            'unmatched_ocr': 1
        }

        formatted = format_metrics_for_display(metrics)

        self.assertEqual(formatted['precision'], "66.67%")
        self.assertEqual(formatted['recall'], "50.00%")
        self.assertEqual(formatted['avg_crr'], "75.00%")
        self.assertIn("57.1", formatted['f1_score'])

    def test_format_preserves_counts(self):
        """Counts should not be formatted"""
        metrics = {
            'precision': 1.0,
            'recall': 1.0,
            'avg_crr': 1.0,
            'f1_score': 1.0,
            'exact_matches': 5,
            'fuzzy_matches': 2,
            'total_gt_words': 7,
            'total_ocr_words': 7,
            'unmatched_gt': 0,
            'unmatched_ocr': 0
        }

        formatted = format_metrics_for_display(metrics)

        self.assertEqual(formatted['exact_matches'], 5)
        self.assertEqual(formatted['fuzzy_matches'], 2)
        self.assertEqual(formatted['total_gt_words'], 7)


if __name__ == '__main__':
    unittest.main()
